types/react/v17/jsx-runtime.d.ts

```diff
@@ -1,2 +1,17 @@
-// Expose `JSX` namespace in `global` namespace
-import './';
+import { ElementType, Fragment, Key, ReactElement } from '.';
+
+export { Fragment };
+
+/**
+ * Create a React element.
+ *
+ * You should not use this function directly. Use JSX and a transpiler instead.
```

Comment:
Why do we need types for them now? It feels a bit scary to publish them without any actual usage or tests. I'd like to avoid people using them since that's not their intended purpose.

Reply:
> However, when inspecting compiled code, these functions show up.

I don't understand how that's relevant to the types. It sounds like you're debugging the transformer plugin and want to use the types to verify? 

The larger point is that type-checking JSX is different than normal type-checking in TS. You can't just transform JSX, type-check and expect the same result. I'd like to avoid giving that impression. Otherwise we end up with types like that for `createElement` which are way more complicated than they need to be.

If TypeScript uses the runtime types for type-checking, we need to rethink their types anyway. I'd rather test this either with their intended design that's going to ship or, if they don't want to collaborate on that, work on it once shipped. 

But building them out before just adds more technical debt.

>  Both implemented it wrong initially.

What mistakes did they make?

Reply:
The use case I need it for is the first in the list. I need to wrap React’s automatic runtime, which requires these type definitions.

> > However, when inspecting compiled code, these functions show up.
> 
> I don't understand how that's relevant to the types.

For example, create a new directory, and install a React library:

```sh
mkdir project
cd project
npm install @mui/material
```

Now open `./node_modules/@mui/material/Accordion/Accordion.js`. It has these `jsx` and `jsxs` variables. TypeScript doesn’t know them, so it shows them as `any`. I think having these types helps users understand a tiny bit better what these functions are, and what code they’re actually running.

> It sounds like you're debugging the transformer plugin and want to use the types to verify?

Not currently, but I have contributed to the JSX transformer in MDX. I see how that could be a potential use case.

> The larger point is that type-checking JSX is different than normal type-checking in TS. You can't just transform JSX, type-check and expect the same result. I'd like to avoid giving that impression. Otherwise we end up with types like that for `createElement` which are way more complicated than they need to be.

I am fully aware how type checking JSX works for both the classic and automatic runtime. I share your sentiment to avoid overcomplication, which is why I kept the types minimal.

> If TypeScript uses the runtime types for type-checking, we need to rethink their types anyway. I'd rather test this either with their intended design that's going to ship or, if they don't want to collaborate on that, work on it once shipped.

I’m not aware any work is being done on that.

> > Both implemented it wrong initially.
>
> What mistakes did they make?

Both basically exported the classic runtime pragma as `jsx`, `jsxs`, and `jsxDEV`. It’s fixed now. I feel this is out of scope for the discussion. My point is these types might help framework authors by providing a reference of how to implement the automatic runtime.

Reply:
> I’m not aware any work is being done on that.

It's scheduled for 5.1 (https://github.com/microsoft/TypeScript/issues/53031) which is why I'm hesitant to work on this now in parallel until we have clarification from the TypeScript team. Which is usually after things have shipped which makes me even more hesitant to continue on this at this time.

There's kind of a chicken-egg problem here where we don't have good JSX factory types because we don't really know how they best look like if used for JSX type-checking while TS is trying to use the JSX factory for JSX type-checking. It would be nice if we'd be more involved so that the TS team is aware of requirements and constraints we have. But I haven't heard back from them so at this point we just have to wait and see.

Reply:
I wasn't aware of that. Then it makes perfect sense to delay this.

Reply:
I’ll close this PR in favor of whatever implementation will be added once TypeScript 5.1 is released. :)